clients/drcachesim/scheduler/scheduler_impl.h

```diff
@@ -976,6 +1000,56 @@ typedef scheduler_impl_tmpl_t<memref_t, reader_t> scheduler_impl_t;
 typedef scheduler_impl_tmpl_t<trace_entry_t, dynamorio::drmemtrace::record_reader_t>
     record_scheduler_impl_t;
 
+template <typename RecordType, typename ReaderType>
+class scheduler_dynamic_tmpl_t : public scheduler_impl_tmpl_t<RecordType, ReaderType> {
+private:
+    using sched_type_t = scheduler_tmpl_t<RecordType, ReaderType>;
+    using input_ordinal_t = typename sched_type_t::input_ordinal_t;
+    using output_ordinal_t = typename sched_type_t::output_ordinal_t;
+    using stream_status_t = typename sched_type_t::stream_status_t;
+    using input_info_t =
+        typename scheduler_impl_tmpl_t<RecordType, ReaderType>::input_info_t;
+    using output_info_t =
+        typename scheduler_impl_tmpl_t<RecordType, ReaderType>::output_info_t;
+    using schedule_record_t =
+        typename scheduler_impl_tmpl_t<RecordType, ReaderType>::schedule_record_t;
+
+protected:
+    stream_status_t
+    pick_next_input_try(output_ordinal_t output, uint64_t blocked_time,
+                        input_ordinal_t prev_index, input_ordinal_t &index) override;
+};
+
+template <typename RecordType, typename ReaderType>
+class scheduler_replay_tmpl_t : public scheduler_impl_tmpl_t<RecordType, ReaderType> {
+private:
+    using sched_type_t = scheduler_tmpl_t<RecordType, ReaderType>;
+    using input_ordinal_t = typename sched_type_t::input_ordinal_t;
```

Comment:
Couldn't we move these to the protected section of scheduler_fixed_tmpl_t to avoid redeclaring in each?

Reply:
I tried it but unfortunately it doesn't work: the compiler complains about unknown types in both the header and .cpp.